File: verify/js/tree.js
/**
 * In-memory tree verify — mirrors hashseal-core tree hash + verify policy.
 * Zero npm dependencies. Used for multi-lang tree-v1 vectors without a filesystem walk.
 *
 * Copyright (c) 2026 MonkeyKing.dev
 */

"use strict";

const { blake3Digest } = require("./check.js");

const DEFAULT_TEXT_EXTENSIONS = new Set([
  "md",
  "txt",
  "toml",
  "yml",
  "yaml",
  "json",
  "rs",
  "java",
  "go",
  "py",
  "js",
  "ts",
  "tsx",
  "jsx",
  "css",
  "html",
  "xml",
  "sh",
  "ps1",
  "c",
  "h",
  "cpp",
  "cs",
  "rb",
  "svg",
]);

function normalizeLf(s) {
  return s.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
}

function extOf(path) {
  const i = path.lastIndexOf(".");
  if (i < 0) return "";
  return path.slice(i + 1).toLowerCase();
}

/**
 * Hash one path+content with core tree policy.
 * @param {string} path posix relative path
 * @param {string} content file body (UTF-8 string; binary as latin1 if needed)
 * @param {{ lineEndingsLfText?: boolean, textExtensions?: Set<string>|string[] }} [opts]
 * @returns {{ digest: string, size: number, qualified: string, hex: string }}
 */
function hashTreeFileContent(path, content, opts) {
  const lfText = !opts || opts.lineEndingsLfText !== false;
  const textExts =
    opts && opts.textExtensions
      ? opts.textExtensions instanceof Set
        ? opts.textExtensions
        : new Set(opts.textExtensions)
      : DEFAULT_TEXT_EXTENSIONS;
  const size =
    typeof Buffer !== "undefined"
      ? Buffer.byteLength(content, "utf8")
      : new TextEncoder().encode(content).length;
  let data = content;
  if (lfText && textExts.has(extOf(path))) {
    data = normalizeLf(content.replace(/^\uFEFF/, ""));
  }
  const d = blake3Digest(data);
  return {
    digest: d.qualified,
    qualified: d.qualified,
    hex: d.hex,
    size,
  };
}

/**
 * Verify in-memory files against ledger entries (same findings as hashseal-core verify_tree).
 * @param {Record<string, string>} files path → content
 * @param {Array<{ path: string, digest: string, size?: number }>} ledgerEntries
 * @param {{ lineEndingsLfText?: boolean, textExtensions?: Set<string>|string[] }} [opts]
 * @returns {{ ok: boolean, checked: number, findings: Array<{path:string,status:string,expected:string|null,actual:string|null}> }}
 */
function verifyTreeInMemory(files, ledgerEntries, opts) {
  const current = new Map();
  const paths = Object.keys(files || {}).sort();
  for (const p of paths) {
    const h = hashTreeFileContent(p, files[p], opts);
    current.set(p, h.qualified);
  }

  const findings = [];
  const expectedPaths = new Set();
  const entries = ledgerEntries || [];

  for (const e of entries) {
    expectedPaths.add(e.path);
    const actual = current.get(e.path);
    if (actual === undefined) {
      findings.push({
        path: e.path,
        status: "removed",
        expected: e.digest,
        actual: null,
      });
    } else if (actual !== e.digest) {
      findings.push({
        path: e.path,
        status: "mismatch",
        expected: e.digest,
        actual,
      });
    }
  }

  for (const [path, digest] of current) {
    if (!expectedPaths.has(path)) {
      findings.push({
        path,
        status: "added",
        expected: null,
        actual: digest,
      });
    }
  }

  findings.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
  return {
    ok: findings.length === 0,
    checked: entries.length,
    findings,
  };
}

module.exports = {
  hashTreeFileContent,
  verifyTreeInMemory,
  DEFAULT_TEXT_EXTENSIONS,
  normalizeLf,
};

File: verify/python/tree.py
"""
In-memory tree verify — mirrors hashseal-core tree hash + verify policy.
Zero pip dependencies. Used for multi-lang tree-v1 vectors without a filesystem walk.

Copyright (c) 2026 MonkeyKing.dev
"""

from __future__ import annotations

from typing import Any, Dict, List, Mapping, Optional, Sequence, Set, Union

from check import blake3_digest

DEFAULT_TEXT_EXTENSIONS: frozenset[str] = frozenset(
    {
        "md",
        "txt",
        "toml",
        "yml",
        "yaml",
        "json",
        "rs",
        "java",
        "go",
        "py",
        "js",
        "ts",
        "tsx",
        "jsx",
        "css",
        "html",
        "xml",
        "sh",
        "ps1",
        "c",
        "h",
        "cpp",
        "cs",
        "rb",
        "svg",
    }
)


def normalize_lf(s: str) -> str:
    return s.replace("\r\n", "\n").replace("\r", "\n")


def ext_of(path: str) -> str:
    i = path.rfind(".")
    if i < 0:
        return ""
    return path[i + 1 :].lower()


def _text_ext_set(
    text_extensions: Optional[Union[Set[str], Sequence[str], frozenset[str]]],
) -> Set[str]:
    if text_extensions is None:
        return set(DEFAULT_TEXT_EXTENSIONS)
    if isinstance(text_extensions, set):
        return text_extensions
    return set(text_extensions)


def hash_tree_file_content(
    path: str,
    content: str,
    *,
    line_endings_lf_text: bool = True,
    text_extensions: Optional[Union[Set[str], Sequence[str], frozenset[str]]] = None,
) -> Dict[str, Any]:
    """Hash one path+content with core tree policy.

    size is on-disk UTF-8 byte length before normalize.
    """
    text_exts = _text_ext_set(text_extensions)
    size = len(content.encode("utf-8"))
    data = content
    if line_endings_lf_text and ext_of(path) in text_exts:
        if data and ord(data[0]) == 0xFEFF:
            data = data[1:]
        data = normalize_lf(data)
    d = blake3_digest(data)
    return {
        "digest": d["qualified"],
        "qualified": d["qualified"],
        "hex": d["hex"],
        "size": size,
    }


def verify_tree_in_memory(
    files: Optional[Mapping[str, str]],
    ledger_entries: Optional[Sequence[Mapping[str, Any]]],
    *,
    line_endings_lf_text: bool = True,
    text_extensions: Optional[Union[Set[str], Sequence[str], frozenset[str]]] = None,
) -> Dict[str, Any]:
    """Verify in-memory files against ledger entries (same findings as hashseal-core verify_tree)."""
    current: Dict[str, str] = {}
    for p in sorted((files or {}).keys()):
        h = hash_tree_file_content(
            p,
            files[p],  # type: ignore[index]
            line_endings_lf_text=line_endings_lf_text,
            text_extensions=text_extensions,
        )
        current[p] = h["qualified"]

    findings: List[Dict[str, Any]] = []
    expected_paths: Set[str] = set()
    entries = list(ledger_entries or [])

    for e in entries:
        path = e["path"]
        expected_paths.add(path)
        actual = current.get(path)
        if actual is None:
            findings.append(
                {
                    "path": path,
                    "status": "removed",
                    "expected": e["digest"],
                    "actual": None,
                }
            )
        elif actual != e["digest"]:
            findings.append(
                {
                    "path": path,
                    "status": "mismatch",
                    "expected": e["digest"],
                    "actual": actual,
                }
            )

    for path, digest in current.items():
        if path not in expected_paths:
            findings.append(
                {
                    "path": path,
                    "status": "added",
                    "expected": None,
                    "actual": digest,
                }
            )

    findings.sort(key=lambda f: f["path"])
    return {
        "ok": len(findings) == 0,
        "checked": len(entries),
        "findings": findings,
    }


# Public aliases matching JS SDK style
hashTreeFileContent = hash_tree_file_content
verifyTreeInMemory = verify_tree_in_memory

__all__ = [
    "DEFAULT_TEXT_EXTENSIONS",
    "normalize_lf",
    "hash_tree_file_content",
    "hashTreeFileContent",
    "verify_tree_in_memory",
    "verifyTreeInMemory",
]
